modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/RegexReplace.java

```diff
@@ -0,0 +1,177 @@
+                try {
+                    Pattern p = Pattern.compile(rule.getAttribute("match"));
+                    PatternReplace pr = new PatternReplace(p, rule.getAttribute("replace"));
+                    patterns.add(pr);
+                    DEBUG(pr);
+                } catch(PatternSyntaxException pse) {
+                    logger.error("There is a problem with the regular expression!");
+                    logger.error("The pattern in question is: "+pse.getPattern());
+                    logger.error("The description is: "+pse.getDescription());
+                    logger.error("The message is: "+pse.getMessage());
+                    logger.error("The index is: "+pse.getIndex());
+
+                    throw new IllegalArgumentException(pse.getMessage());
+                }
+            }
+        } catch (ParserConfigurationException | SAXException | IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Evaluates an XPath expression and returns the result as a {@link NodeList}.
+     * @param node the {@link Node} to evaluate the expression on
+     * @param xpath the XPath expression to evaluate
+     * @return a {@link NodeList}, or <code>null</code> if there is an
+     * error in the XPath expression.
+     */
+    private NodeList selectNodes(Node node, String xpath) {
+        try {
+            return (NodeList) this.xpathObj.evaluate(xpath, node, XPathConstants.NODESET);
+        } catch (XPathExpressionException e) {
+            return null;
+        }
+    }
+
+    /**
+     * Filters a string using the regexes.
+     * @param text the text to filter.
+     * @return the filtered text.
+     */
+    public String filter(String text) {
+        for (Iterator<PatternReplace> it = patterns.iterator(); it.hasNext(); ) {
+            PatternReplace pr = it.next();
+            text = processTest(text, pr.getPattern(), pr.getReplace());
+        }
+        return text;
+    }
+
+    /**
+     * Performes a regex-replace.
+     * @param input the input text.
+     * @param pattern the pattern to match.
+     * @param replace the replacement string.
+     * @return the filtered text.
+     */
+    private String processTest(String input, Pattern pattern, String replace) {
+        try {
+            Matcher matcher = pattern.matcher(input);
+            return matcher.replaceAll(replace);
+        } catch (Exception e) {
+            logger.error("RegexReplace:");
+            logger.error("     pattern: " + pattern.toString());
+            logger.error("       input: " + input);
+            logger.error("     replace: " + replace);
+
+            throw new IllegalArgumentException(e.getMessage());
+        }
+    }
+
+
+    private void DEBUG(PatternReplace pr) {
+        DEBUG(pr.getPattern().toString() + "\t" + pr.getReplace());
+    }
+
+
+    private void DEBUG(String msg) {
+        if (DEBUG) {
+            logger.error("RegexReplace: " + msg);
```

Comment:
Shouldn't this be logger.debug()?

Reply:
Oop, yes. To much copy paste there i think. I will fix it.

Reply:
Updated